### ISSUE TITLE
Replace HF Transformers Whisper + pyannote with WhisperX pipeline

### ISSUE THREAD

**Issue by brlauuu** [brlauuu/podlog#7]
## Summary

Replace the current two-step transcription pipeline (HuggingFace Transformers Whisper → pyannote diarization) with [WhisperX](https://github.com/m-bain/whisperX), which bundles faster-whisper (CTranslate2) + wav2vec2 word-level alignment + pyannote diarization into a single library.

**Primary motivation:** ~4x faster CPU transcription via CTranslate2 backend with int8 quantization, plus word-level timestamps for free.

## Background

The current pipeline uses:
- `transformers` `AutoModelForSpeechSeq2Seq` + HF `pipeline()` for Whisper transcription (very slow on CPU)
- `pyannote-audio` `speaker-diarization-3.1` for diarization (separate Celery task)
- A custom `alignment.py` service for majority-overlap speaker assignment

WhisperX wraps faster-whisper (CTranslate2) as its ASR backend, adds wav2vec2-based forced phoneme alignment for word-level timestamps, and uses pyannote for diarization — the same pyannote pipeline we already use.

### 1. Dependencies (`pyproject.toml`)

**Remove:**
```
transformers = "^4.40.0"
accelerate = "^0.29.0"
```

**Add:**
```
whisperx = "^3.8.0"
```

WhisperX pulls in `faster-whisper`, `ctranslate2`, `pyannote-audio`, `torch`, `torchaudio`, and `transformers` as transitive deps. The net effect is: we drop the explicit `transformers` + `accelerate` pins and gain `faster-whisper`/`ctranslate2`/`torchaudio` via whisperx.

`pyannote-audio` and `torch` remain (WhisperX depends on both). `ffmpeg-python` is still needed for the WAV conversion step in `transcribe.py`.

After updating `pyproject.toml`, regenerate `poetry.lock`.

### 2. Config (`apps/pipeline/app/config.py`)

**Add new settings:**
```python
# Whisper
whisper_model: str = "large-v3-turbo"   # existing, keep as-is
whisper_backend: str = "whisperx"       # new: "whisperx" or "transformers"
whisper_compute_type: str = "int8"      # new: CTranslate2 compute type for CPU
whisper_batch_size: int = 16            # new: batch size for WhisperX batched inference
whisper_vad_filter: bool = True         # new: enable Silero VAD silence removal
```

Also update `.env.example` with the new vars and comments.

> **Note on large-v3-turbo support:** WhisperX does not officially list `large-v3-turbo` as a supported model string. It works via the community CTranslate2 conversion at `deepdml/faster-whisper-large-v3-turbo-ct2`. Test this during implementation — if it doesn't work with the plain string, use the full HF repo ID. Alternatively, fall back to `large-v3` which is fully supported. See [m-bain/whisperX#898](https://github.com/m-bain/whisperX/issues/898).

### 3. Whisper service (`apps/pipeline/app/services/whisper.py`)

Rewrite to use WhisperX. The public API (`load_model`, `unload_model`, `transcribe`) must keep the same signatures and return types so the Celery tasks don't need to change their contract.

**Current signatures to preserve:**
```python
def load_model(model_name: str = "large-v3-turbo") -> None
def unload_model() -> None
def transcribe(audio_path: str, model_name: str = "large-v3-turbo") -> tuple[list[dict], str]
```

**Current return format for `transcribe()`:**
```python
# segments: [{"start": float, "end": float, "text": str}, ...]
# language: str (e.g. "en" or "unknown")
```

**New implementation:**
```python
import gc
import logging
from typing import Optional

logger = logging.getLogger(__name__)

_model = None

def load_model(model_name: str = "large-v3-turbo") -> None:
    global _model
    if _model is not None:
        return

    import whisperx
    from app.config import settings

    device = "cuda" if _cuda_available() else "cpu"
    compute_type = settings.whisper_compute_type  # "int8" for CPU

    logger.info('"action": "whisper_load_start", "model": "%s", "backend": "whisperx"', model_name)

    _model = whisperx.load_model(
        model_name,
        device=device,
        compute_type=compute_type,
    )

    logger.info('"action": "whisper_load_complete", "model": "%s", "device": "%s"', model_name, device)


def unload_model() -> None:
    global _model
    _model = None
    try:
        import torch
        torch.cuda.empty_cache()
    except Exception:
        pass
    gc.collect()
    logger.info('"action": "whisper_unloaded"')


def transcribe(audio_path: str, model_name: str = "large-v3-turbo") -> tuple[list[dict], str]:
    import whisperx
    from app.config import settings

    load_model(model_name)

    device = "cuda" if _cuda_available() else "cpu"
    audio = whisperx.load_audio(audio_path)

    # Stage 1: Transcribe (batched inference via faster-whisper)
    result = _model.transcribe(audio, batch_size=settings.whisper_batch_size)
    language = result.get("language", "unknown") or "unknown"

    # Stage 2: Word-level alignment (optional but recommended)
    try:
        model_a, metadata = whisperx.load_align_model(language_code=language, device=device)
        result = whisperx.align(result["segments"], model_a, metadata, audio, device)
        del model_a  # free alignment model memory
        gc.collect()
    except Exception as exc:
        logger.warning('"action": "whisper_align_failed", "error": "%s"', exc)
        # Fall back to segment-level timestamps (still usable)

    # Convert to expected output format
    segments = []
    for seg in result.get("segments", []):
        segments.append({
            "start": seg.get("start", 0.0),
            "end": seg.get("end", 0.0),
            "text": seg.get("text", "").strip(),
        })

    logger.info(
        '"action": "whisper_transcribe_complete", "segments": %d, "language": "%s"',
        len(segments), language,
    )
    return segments, language


def _cuda_available() -> bool:
    try:
        import torch
        return torch.cuda.is_available()
    except Exception:
        return False
```

**Key differences from current implementation:**
- Uses `whisperx.load_model()` instead of `AutoModelForSpeechSeq2Seq` + `AutoProcessor`
- Uses `whisperx.load_audio()` + `model.transcribe()` instead of HF `pipeline()`
- Adds optional wav2vec2 alignment stage (graceful failure — falls back to segment-level)
- `compute_type="int8"` enables CTranslate2 quantization (big CPU speedup)
- No `_processor` global needed (WhisperX manages internally)

### 4. Diarization option: WhisperX diarization vs current pyannote

WhisperX's diarization is the same pyannote pipeline we already use. Two approaches:

**Option A (recommended): Keep diarization as-is.** The current `pyannote.py` service and `diarize.py` task work correctly. WhisperX's diarization adds no quality improvement — it's the same pyannote/speaker-diarization-3.1. The only advantage would be `assign_word_speakers()` which can assign speakers at word-level granularity, but our segment-level majority-overlap approach in `alignment.py` is sufficient for podcast search.

**Option B: Use WhisperX diarization.** Replace `pyannote.py` and `alignment.py` with WhisperX's built-in diarize + assign pipeline. This simplifies code but couples us more tightly to WhisperX. Implementation:

```python
# In diarize.py task, replace pyannote.diarize() + alignment.assign_speakers() with:
import whisperx
from whisperx.diarize import DiarizationPipeline

diarize_model = DiarizationPipeline(use_auth_token=settings.hf_token, device=device)
diarize_segments = diarize_model(audio)
result = whisperx.assign_word_speakers(diarize_segments, transcript_result)
```

If going with Option B, the diarize task would need access to the full WhisperX result dict (with word-level alignment), not just the segment list from the DB. This would require either:
- Storing the WhisperX result as JSON between tasks, or
- Merging transcribe + diarize into a single Celery task

**Recommendation:** Go with Option A for the initial switch. It's lower risk and preserves the existing task separation.

### 5. Prewarm task (`apps/pipeline/app/tasks/prewarm.py`)

Update to use WhisperX model loading:

```python
# Current:
from app.services.whisper import load_model, unload_model
load_model(settings.whisper_model)
unload_model()

# No change needed — the whisper.py service's load_model/unload_model
# interface is preserved. The prewarm task calls these unchanged.
```

The prewarm will now download the CTranslate2-format model instead of the PyTorch model. First run will take longer as it downloads the converted weights.

### 6. Transcribe task (`apps/pipeline/app/tasks/transcribe.py`)

**No changes needed** to the task itself. It calls:
- `whisper.transcribe(str(wav_path), model_name=settings.whisper_model)` → same signature, same return type
- `whisper.unload_model()` via `_unload_whisper()` → same interface

The `_unload_whisper()` helper at the bottom of `transcribe.py` directly accesses `whisper_mod._model`. This still works since the new implementation keeps `_model` as a module-level global. However, verify that `_unload_whisper()` cleans up properly:

```python
# Current _unload_whisper() in transcribe.py:
def _unload_whisper() -> None:
    import sys
    whisper_mod = sys.modules.get("app.services.whisper")
    if whisper_mod and hasattr(whisper_mod, "_model"):
        whisper_mod._model = None
    try:
        import torch
        torch.cuda.empty_cache()
    except Exception:
        pass
    gc.collect()
```

This should still work since the new `whisper.py` keeps `_model` as the module global. No `_processor` to clear (WhisperX manages that internally).

### 7. Diarize task (`apps/pipeline/app/tasks/diarize.py`)

**No changes needed** if going with Option A (keep pyannote as-is).

### 8. Docker (`apps/pipeline/Dockerfile`)

- `ffmpeg` system dependency is still needed (WAV conversion in transcribe task)
- No new system dependencies required (WhisperX bundles PyAV for its own audio loading)
- The Docker image will be slightly larger due to `ctranslate2` and `torchaudio` additions

### 9. Tests

**Unit tests to update:**
- `tests/unit/test_whisper_service.py` — mock `whisperx.load_model` and `model.transcribe` instead of `transformers.AutoModelForSpeechSeq2Seq` and `transformers.pipeline`
- `tests/unit/test_transcribe_task.py` — mock paths change from `app.services.whisper.transcribe` (same name but verify mock targets)

**New test cases to add:**
- Verify `transcribe()` returns correct format when alignment succeeds
- Verify `transcribe()` returns correct format when alignment fails (graceful fallback)
- Verify `compute_type` and `batch_size` are passed through from settings
- Verify `unload_model()` cleans up properly (no residual WhisperX objects)

**Integration tests:**
- Test with a real audio file to verify the CTranslate2 model loads and produces output
- Test that the prewarm downloads the correct model format

### 10. Model cache

The HuggingFace cache volume (`/root/.cache/huggingface`) is shared between pipeline, worker, and beat containers. WhisperX/faster-whisper downloads CTranslate2-format models to `~/.cache/huggingface/hub/` (same location, different model files than PyTorch format).

**After switching:** The old PyTorch Whisper model files will remain in the cache but won't be used. They can be cleaned up manually to reclaim disk space (~6GB for large-v3).

## Risks and Mitigations

| Risk | Mitigation |
|---|---|
| `large-v3-turbo` not supported by WhisperX | Fall back to `large-v3` or use HF repo ID `deepdml/faster-whisper-large-v3-turbo-ct2`. Test during implementation. |
| Dependency conflicts between whisperx, pyannote, torch | Pin versions carefully. WhisperX 3.8+ bundles compatible pyannote. Run `poetry lock` and verify. |
| WhisperX segments generator gotcha | WhisperX's `model.transcribe()` returns a dict with materialized segments (not a generator like raw faster-whisper). No issue. |
| int8 quantization accuracy regression | Benchmark a few episodes against current float32 output. <2% degradation expected per CTranslate2 docs. |
| Alignment model download on first run | Prewarm task should also download the wav2vec2 alignment model. Add a `whisperx.load_align_model()` call to prewarm. |
| Memory: alignment model loaded alongside whisper | wav2vec2 alignment models are small (~1GB). Monitor peak RSS. If too high, skip alignment. |

## Acceptance Criteria

- [ ] `whisperx` replaces `transformers` pipeline for transcription
- [ ] `compute_type="int8"` used on CPU for performance
- [ ] `transcribe()` returns same `(segments, language)` format — no downstream changes needed
- [ ] Word-level alignment runs when available, gracefully skipped on failure
- [ ] Existing pyannote diarization continues to work unchanged (Option A)
- [ ] Prewarm downloads CTranslate2 model + alignment model
- [ ] All existing unit tests pass (with updated mocks)
- [ ] New config vars (`whisper_compute_type`, `whisper_batch_size`) documented in `.env.example`
- [ ] Docker image builds successfully
- [ ] Manual test: full pipeline processes an episode end-to-end

## Out of Scope

- Replacing pyannote diarization with WhisperX diarization (Option B — separate issue if desired)
- Word-level timestamps in the `segments` DB table (would require schema change)
- GPU/CUDA optimizations (CPU-only deployment)
